lib/cn/csched_sp3.c

```diff
@@ -2144,48 +2166,45 @@ struct periodic_check {
 static void
 sp3_monitor(struct work_struct *work)
 {
-    struct sp3 *sp = container_of(work, struct sp3, wstruct);
-
-    const int timeout_ms = 100;
-    bool      bad_health = false;
-
-    struct periodic_check chk_qos;
-    struct periodic_check chk_refresh;
-    struct periodic_check chk_shape;
+    struct sp3 *sp = container_of(work, struct sp3, mon_work);
 
-    u64 now, last_activity;
+    struct periodic_check chk_qos     = { .interval = NSEC_PER_SEC / 5 };
+    struct periodic_check chk_sched   = { .interval = NSEC_PER_SEC };
+    struct periodic_check chk_refresh = { .interval = NSEC_PER_SEC * 10 };
+    struct periodic_check chk_shape   = { .interval = NSEC_PER_SEC * 15 };
 
-    now = get_time_ns();
-    last_activity = now;
-
-    chk_qos.interval = NSEC_PER_SEC / 5;
-    chk_refresh.interval = 10 * NSEC_PER_SEC;
-    chk_shape.interval = 15 * NSEC_PER_SEC;
-
-    chk_qos.next = now + chk_qos.interval;
-    chk_refresh.next = now + chk_refresh.interval;
-    chk_shape.next = now + chk_shape.interval;
+    bool bad_health = false;
+    u64 last_activity = 0;
 
     sp3_refresh_settings(sp);
 
-    while (!atomic_read(&sp->destruct)) {
+    while (atomic_read(&sp->running)) {
+        uint64_t now = get_time_ns();
         merr_t err;
 
-        if (!sp->activity || bad_health) {
-            mutex_lock(&sp->mutex);
-            cv_timedwait(&sp->cv, &sp->mutex, timeout_ms);
-            mutex_unlock(&sp->mutex);
-        }
+        mutex_lock(&sp->mon_lock);
+        if (!sp->mon_signaled && now < chk_qos.next) {
+            int timeout_ms = max_t(int, 10, (chk_qos.next - now) / USEC_PER_SEC);
 
-        now = get_time_ns();
+            cv_timedwait(&sp->mon_cv, &sp->mon_lock, timeout_ms);
 
-        sp->activity = 0;
+            now = get_time_ns();
+        }
+        sp->mon_signaled = false;
+        mutex_unlock(&sp->mon_lock);
```

Comment:
sp->mon_signaled should make this more responsive to external calls to sp3_monitor_wake(), and we should now hit the qos check deadline (and other chk deadlines) more periodically than before...